### PR TITLE
Edit Products: update product name edit UI to inline text field

### LINKED ISSUE #1957: Products: Title field is huge/not edited inline
**Describe the bug**

According to the designs, the title field should be edited inline:

image

However, the current implementation has the title field use a full-screen editor (same as the description field), encouraging very long, multi-line text.

**To Reproduce**
1. Go to the Products tab.
2. Select a product.
3. Select "Title."

**Screenshots**

**Mobile Environment**
Please include:
- Device: iPhone XS
- iOS version: 13.3.1
- WooCommerce iOS version: 3.8.0.0 TestFlight beta

### PR DESCRIPTION
**Pull request by jaclync**
Fixes #1957 

## Changes

- Added `UITextField` extension to apply headline styles
- Created `TextFieldTableViewCell` that contains a text field + xib
- Updated `ProductFormTableViewDataSource` and `ProductFormSection.PrimaryFieldRow` configuration to use the new `TextFieldTableViewCell`
- In `ProductFormViewController` and `ProductFormTableViewDataSource`, updated the name change handling. When the name changes, it still updates its `product` but it won't trigger a table view reload if name is the only change

## Testing

- Go to the Products tab
- Tap on a simple Product
- Edit the product name however you like. Try clearing the title to make sure the placeholder says "Title"
- Tap on any other row, change some data there, and tap "Done" --> it should go back to the Edit Product main screen, and the product name should remain the same as the last edited value
- Tap "Update" --> the product name should be updated remotely after it completes

## Example screenshots

Screencast:

simulator

mode \ state | empty | non-empty
-- | -- | --
dark | Simulator Screen Shot - iPhone 11 - 2020-03-20 at 18 06 18 | Simulator Screen Shot - iPhone 11 - 2020-03-20 at 18 06 20
light | Simulator Screen Shot - iPhone 11 - 2020-03-20 at 18 05 37 | Simulator Screen Shot - iPhone 11 - 2020-03-20 at 18 05 40

 
Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift [hunk 1]
```diff
@@ -220,8 +241,7 @@ extension ProductFormViewController: UITableViewDelegate {
             case .images:
                 break
             case .name:
-                ServiceLocator.analytics.track(.productDetailViewProductNameTapped)
-                editProductName()
+                break
```
[the remark this answers is not in the document]
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift]: yup! updated in https://github.com/woocommerce/woocommerce-ios/pull/2012/commits/b3df31ba3ae70aa5263eece62dfaaa3eb610d213

**Comment by jaclync** [comment]
Thanks for the thorough testing/review @pmusolino ! I updated the PR for another look, and responded to your questions below:

> 1. It seems that we removed the event productDetailViewProductNameTapped, so we don't know when the title was tapped. Maybe we can fire the event when the textfield will be selected.

 great catch! I added back the event and tracked it when the product name text field did begin editing in: https://github.com/woocommerce/woocommerce-ios/pull/2012/commits/bb49387d59199f767d109b663bb508127eb7de91

> 2. It seems that we also removed the event productNameDoneButtonTapped, so we don't know when the title will be updated.

yes since we don't have a button to commit the product name update anymore, I've removed this event in https://github.com/woocommerce/woocommerce-ios/pull/2012/commits/8f1747b96394118bbdeec05e2c8d9330133fe384. Android also doesn't track this event.

> 3. This is just a UX improvement: the first thing that I did after finishing to type the new title was to press return for closing the keyboard, but nothing happens. Maybe it can be useful to close the keyboard, because at the moment the only way to close it is to update the product or tap another cell.

great idea, I think it'd be nice to dismiss the keyboard when the user taps return on the keyboard. implemented in: https://github.com/woocommerce/woocommerce-ios/pull/2012/commits/d0ab01e6c17bb0cdf15a3822591de1bf52bc1197

> 4. Do you think that it make sense to have unit tests for TextFieldTableViewCell since for other cells we wrote the tests?

I'm a bit torn on this one -  also tagging @shiki for discussion: we have a few pre-existing unit tests on `UITableViewCell` subclasses, like [`StatusListTableViewCellTests`](https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/WooCommerceTests/ViewRelated/StatusListTableViewCellTests.swift). While it's nice to make sure individual UI elements in the cell are matching design, I have some concerns:
- (major) these subviews that are being tested cannot be private and could be called externally by accident outside of unit tests.
- right now we have a bunch of asserts on label subviews by making a `UILabel` and calling `label.apply*Style()` ([example](https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/WooCommerceTests/ViewRelated/StatusListTableViewCellTests.swift#L19-L20)). we also talked about converting `apply*Style()` extensions to view subclasses, and we'll have to update all the unit tests that use this type of custom style subclasses. this is fine as part of the refactoring, somehow it just felt like lots of overhead whenever we make changes to the UI layer.

I'm open to opening up the subviews for unit tests though, just wanted to hear what the team thinks!

> 5. If you tap near the border of the cell, I can select the cell but not the text field. Maybe extending the size of the text field (top and bottom padding = 0) can solve the problem.

nice catch, the highlighted state doesn't look good. I didn't find a native API to set insets to a `UITextField`, and to keep things simple for now I just removed the highlighted state in https://github.com/woocommerce/woocommerce-ios/pull/2012/commits/b3df31ba3ae70aa5263eece62dfaaa3eb610d213

**Comment by shiki** [comment]
Thanks for the ping! I only glanced at `TextFieldTableViewCell`. 

> (major) these subviews that are being tested cannot be private and could be called externally by accident outside of unit tests.

Indeed. I've been struggling with this too. I really don't want to expose those properties just so they can be tested. 

I've taken a balanced approach to this. Most of the time, I can move most of the logic to the `ViewModel` and test the `ViewModel` heavily instead. This can work well especially if we are using something like RxSwift and the view just "reacts" to changes. This has usually been enough for me that I don't add tests for the UI. 

However, there are UIs that are quite complicated that they deserve to have unit tests. It's in this case that I expose those properties. 🙈If we ever have to do this, we could add an [annotation](https://docs.swift.org/swift-book/ReferenceManual/Attributes.html) like [`@VisibleForTesting`](https://developer.android.com/reference/android/support/annotation/VisibleForTesting) to communicate the intention that the property should be private but we have to test it. The annotation is just informational and _should_ be more visible than a comment. 

In the case of `TextFieldTableViewCell`, I'd say it's not that complex at the moment that it should have a unit test so I'm fine with leaving it as is. 

For [`StatusListTableViewCellTests`](https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/WooCommerceTests/ViewRelated/StatusListTableViewCellTests.swift), I don't agree with testing the `font` and `textColor`. Especially in the way that it's tested right now. I believe you can even change the logic in `applyBodyStyle` and the test would still pass. 🤷‍♂ 

> right now we have a bunch of asserts on label subviews by making a UILabel and calling label.apply*Style() (example).

Yup. I don't believe testing these is a good use of our time. I wish we could test more time on testing complex logic instead -- like state management in ViewControllers.

**Comment by jkmassel** [comment]
Howdy – we're freezing `3.9` today, so this PR is being bumped to `4.0`. If needs to be released as part of `4.0`, please re-target it at the `release/4.0` branch.

**Comment by jaclync** [comment]
> You just need to change the border style of the text field (from the cell xib) to the first option, then you can change the top/bottom inset. 👍

hmm I didn't see an option to change the insets of the text field so that I can pin the text field to its superview without the insets 🤔adding insets to a text field might require subclassing or other ways like in [this SO post](https://stackoverflow.com/questions/2694411/text-inset-for-uitextfield). I think it's fine for the 16px horizontal / 10px vertical margin area not tappable for now, and we can support it later if it is a common problem

Screen Shot 2020-03-24 at 10 18 04 AM

**Comment by jaclync** [comment]
re cell UI testing: thanks for the feedback @shiki @pmusolino ! I also agree that is better to test complex logic in the view model, and test the UI properties for more complicated UI.

adding an annotation like `@VisibleForTesting` sounds nice so that we don't have to search the codebase to check if it's only accessed only in tests. let's try this the next time we test UI properties! it'd be cool if we could define some custom scope for a property getter based on the annotation, but it doesn't seem possible now

**Comment by pmusolino** [comment]
I'm sorry @jaclync, I badly explained myself. For "insets" I mean the distance between the border of UITextField and the cell, then the top/bottom constraint 👍

**Pull request merged**